pkgs/build-support/coq/meta-fetch/default.nix

```diff
@@ -0,0 +1,78 @@
+{ lib, fetchFromGitHub }@args:
+let lib = import ../extra-lib.nix {inherit (args) lib;}; in
+with builtins; with lib;
+let
+  isGitHubDomain = d: match "^github.*" d != null;
+  isGitLabDomain = d: match "^gitlab.*" d != null;
+  default-fetcher = {domain ? "github.com", owner ? "", repo, rev, name ? "source", sha256 ? null, ...}@args:
+    let pr = match "^#(.*)$" rev;
+        github-fetch = switch-if [
+        { cond = args?sha256;
+          out = fetchFromGitHub {
+            githubBase = domain;
+            inherit owner repo sha256 rev name;
+          };}
+        { cond = pr != null;
+          out = fetchTarball ({
+            url = "https://api.${domain}/repos/${owner}/${repo}/tarball/pull/${head pr}/head";
+          } // optionalAttrs (args?sha256) { inherit sha256; });
+        }] # default
+        (fetchTarball {
+          url = "https://${domain}/${owner}/${repo}/archive/${rev}.tar.gz";
+            inherit name;
+        });
+        gitlab-fetch = fetchTarball ({
```

Comment:
What about using `fetchzip` or `fetchFromGitLab` when the hash is known?

Reply:
I tried and had some trouble.
I will try again now the rest works.

Reply:
Please do, as it seems to break evaluation (see below the details of grahamcofborg-eval): https://nixos.org/manual/nix/stable/#builtin-fetchTarball.

Reply:
Note that these tarballs may also change:
> hash mismatch in file downloaded from 'https://gitlab.inria.fr/flocq/flocq/-/archive/flocq-3.3.1/flocq-flocq-3.3.1.tar.gz':
  wanted: sha256:0k1nfgiszmai5dihhpfa5mgq9rwigl0n38dw10jn79x89xbdpyh5
  got:    sha256:1mk8adhi5hrllsr0hamzk91vf2405sjr4lh5brg9201mcw11abkz

Reply:
> Please do, as it seems to break evaluation (see below the details of grahamcofborg-eval): https://nixos.org/manual/nix/stable/#builtin-fetchTarball.

Since the last rebase on master I started having trouble with `fetchFromGitHub` as well :cry:
I am switching everything to fetchzip, I hope this will be alright for grahamcofborg-eval...

Reply:
> Note that these tarballs may also change:
> 
> > hash mismatch in file downloaded from 'https://gitlab.inria.fr/flocq/flocq/-/archive/flocq-3.3.1/flocq-flocq-3.3.1.tar.gz':
> > wanted: sha256:0k1nfgiszmai5dihhpfa5mgq9rwigl0n38dw10jn79x89xbdpyh5
> > got:    sha256:1mk8adhi5hrllsr0hamzk91vf2405sjr4lh5brg9201mcw11abkz

I had this as well. I thoroughly evaluated all coqPackages before submitting this PR though, so I am extremely surprised these mismatch suddenly show up :thinking: